kernel/src/schema/mod.rs

```diff
@@ -454,7 +691,53 @@ impl StructType {
         let fields = self
             .fields()
             .map(|field| field.make_physical(column_mapping_mode));
-        Self::new(fields)
+        Self::new_unchecked(fields)
+    }
+
+    /// Validates that there are no metadata columns in the given fields.
+    fn ensure_no_metadata_columns<'a>(
+        fields: impl IntoIterator<Item = &'a StructField>,
+    ) -> DeltaResult<()> {
+        for field in fields {
+            if field.is_metadata_column() {
+                return Err(Error::schema(
+                    "Metadata columns are only allowed at the top level of a schema".to_string(),
+                ));
+            }
+
+            match &field.data_type {
+                DataType::Struct(struct_type) => {
+                    // Only check direct fields, not nested structs (they were validated at creation)
+                    // NB: We collect fields here to prevent infinite recursion errors from the compiler
+                    let fields = struct_type
+                        .fields()
+                        .filter(|f| {
+                            !matches!(f.data_type, DataType::Struct(_) | DataType::Variant(_))
+                        })
+                        .collect::<Vec<_>>();
+                    Self::ensure_no_metadata_columns(fields)?;
+                }
+                DataType::Array(array_type) => {
+                    if let DataType::Struct(struct_type) = array_type.element_type() {
+                        Self::ensure_no_metadata_columns(struct_type.fields())?;
+                    }
+                }
+                DataType::Map(map_type) => {
+                    if let DataType::Struct(struct_type) = map_type.key_type() {
+                        Self::ensure_no_metadata_columns(struct_type.fields())?;
+                    }
+                    if let DataType::Struct(struct_type) = map_type.value_type() {
+                        Self::ensure_no_metadata_columns(struct_type.fields())?;
+                    }
+                }
+                DataType::Variant(variant_type) => {
+                    // Structs inside a variant are not validated at creation, so we need to check them here
+                    Self::ensure_no_metadata_columns(variant_type.fields())?;
+                }
+                _ => {} // Other types cannot contain metadata columns
```

Comment:
To avoid silent breakage as future types are added, recommend to enumerate them here. 
I _think_ the only remaining enum variant is `Primitive`?
```suggestion
                // Other types cannot contain nested metadata columns
                DataType::Primitive(_) => {}
```